Source/WebGPU/WGSL/EntryPointRewriter.cpp

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "EntryPointRewriter.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+
+namespace WGSL {
+
+class EntryPointRewriter {
+public:
+    EntryPointRewriter(AST::ShaderModule& shaderModule, AST::FunctionDecl& functionDecl)
+        : m_shaderModule(shaderModule)
+        , m_functionDecl(functionDecl)
+        , m_emptySourceSpan(0, 0, 0, 0)
+    {
+    }
+
+    void rewrite();
+
+private:
+    struct MemberOrParameter {
+        String m_name;
+        Ref<AST::TypeDecl> m_type;
+        AST::Attribute::List m_attributes;
+    };
+
+    enum class IsBuiltin {
+        No = 0,
+        Yes = 1,
+    };
+
+    static AST::TypeDecl& getResolvedType(AST::TypeDecl&);
+
+    void collectParameters();
+    void constructInputStruct();
+    void materialize(Vector<String>& path, MemberOrParameter&, IsBuiltin);
+    void visit(Vector<String>& path, MemberOrParameter&&);
+    void appendBuiltins();
+
+    AST::ShaderModule& m_shaderModule;
+    AST::FunctionDecl& m_functionDecl;
+
+    const SourceSpan m_emptySourceSpan;
+    Vector<MemberOrParameter> m_builtins;
+    Vector<MemberOrParameter> m_parameters;
+    AST::Statement::List m_materializations;
+    String m_structTypeName;
+    String m_structParameterName;
+};
+
+AST::TypeDecl& EntryPointRewriter::getResolvedType(AST::TypeDecl& type)
```

Comment:
I'm kind of uninformed, but does this work for types that are trees of other types? Like array<f32, 5>?

Reply:
This function should recursively resolve those eventually, e.g.

```
type v = vec;
type f = f32;
v<f, 5>
```

should resolve to `vec<f32, 5>`